### PR TITLE
Post proposals 2

### PR DESCRIPTION
**Pull request by bickelj**
Add a script that POSTs proposal data from CSV.

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -1,10 +1,14 @@
...
   "extends": "@tsconfig/node18/tsconfig.json",
   "compilerOptions": {
-    "noEmit": true,
+    "module": "ES2022",
+    "esModuleInterop": true,
     "resolveJsonModule": true,
     "noUncheckedIndexedAccess": true,
   },
+  "ts-node": {
+    "esm": true,
```
Comment (bickelj) [inline tsconfig.json:10]: I frankly don't recall why I needed to use esm here but I remember it fixing a roadblock.

**Comment by jasonaowen** [comment]
Are you looking for help with the other linter errors, @bickelj? Or is this still a draft because you're intending to fix them and haven't yet?

File: tsconfig.json (hunk 1 above)
Comment (bickelj) [reply tsconfig.json:10]: @jasonaowen Let me try again with cjs without and see what happens.

**Comment by bickelj** [comment]
Yes, help with these linter errors would be appreciated. The reason for the draft status is that there are known lint complaints that may or may not be valid.

File: tsconfig.json (hunk 1 above)
Comment (bickelj) [reply tsconfig.json:10]: I think the problems are much deeper than just cjs vs esm. Scripts don't seem to run on main, for example.

File: src/postProposalVersions.ts [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+      (p) => (p.opportunityId === opportunityId
+        && p.applicantId === applicant.id
+        && p.externalId === proposalExternalId
+      ),
+    );
+  }
+
+  // If the proposal did not exist, create it.
+  if (proposal === undefined) {
+    console.log(`No existing proposal for opportunityId=${opportunityId}, applicantId=${applicant.id}, and externalId=${proposalExternalId}`);
```
Comment (bickelj) [inline src/postProposalVersions.ts:174]: Obviously these console statements can be replaced after a logger is present. These are not the issue.

File: src/postProposalVersions.ts [hunk 3]
```diff
@@ -0,0 +1,320 @@
...
+
+    console.log(`Got ${proposals.length} existing proposals.`);
+
+    const csvParser = csvParse({
+      columns: true,
+    });
+    csvInput.pipe(csvParser);
+
+    // Type of `row` is still `any` here:
+    for await (const csvRow of csvParser) {
```
Comment (bickelj) [inline src/postProposalVersions.ts:275]: Here I am using your suggestion "in particular its async iterator api: https://csv.js.org/parse/api/async_iterator/" which when reading the code is much more elegant. Unfortunately, airbnb disagrees?

File: tsconfig.json (hunk 1 above)
Comment (jasonaowen) [reply tsconfig.json]: That sounds like a big problem! Can you say more? Should that be in a separate issue? When did it stop working?

File: src/postProposalVersions.ts (hunk 3 above)
Comment (jasonaowen) [reply src/postProposalVersions.ts:284]: Yeah, I gave you bad advice - sorry about that! I recently went through something [similar](https://github.com/PhilanthropyDataCommons/data-scripts/pull/24#discussion_r1206113952).

We have time set aside tomorrow to pair on fixing this (and other things), but right now I'd suggest switching over to the [sync version](https://csv.js.org/parse/api/sync/).

File: tsconfig.json (hunk 1 above)
Comment (bickelj) [reply tsconfig.json]: I think this was a local issue, either out of date node modules (run `npm ci`) or something like that. It works fine on main now.

File: src/postProposalVersions.ts (hunk 3 above)
Comment (bickelj) [reply src/postProposalVersions.ts]: After fixing a couple of things we set up the sync version. Instead of an async iterator for loop, we used `csvParser.map` wrapped in a `Promise.all` to allow for use of `async/await` throughout. Many thanks, Jason!

File: src/postProposalVersions.ts (hunk 2 above)
Comment (bickelj) [reply src/postProposalVersions.ts:183]: After @jasonaowen merges a minimal logger setup, I will update this PR with logger statements instead of `console.log`.

Comment (jasonaowen) [reply src/postProposalVersions.ts:183]: See PR #48.

File: src/postProposalVersions.ts [hunk 4]
```diff
@@ -0,0 +1,327 @@
...
+  apiUrl: string;
+}
+
+type CsvRow = Record<string, string>;
+
+function assertIsCsvRow(row: unknown): asserts row is CsvRow {
+  if (row === null
+    || typeof row !== 'object'
+    || Object.keys(row).length === 0
+    || Object.values(row).length === 0) {
```
Comment (jasonaowen) [inline src/postProposalVersions.ts:23]: I kind of alluded to this during our call, but I hadn't fully thought this type guard through at that time: I think we want to be a little bit stricter in our checks here, because we're telling TypeScript some things we're actually not quite sure of. In particular, we want to make sure all the keys and values are actually `string`s, which we're not checking. And, on the other hand, as [`null` is a distinct type and not a universal subtype](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#null-and-undefined), we don't need to both check for `null`ness and check the `typeof` - the latter covers both. Likewise, the API of `Object` is that every key has a value, and vice versa; we don't have to check that there are a nonzero number of both.

```suggestion
type CsvRow = Record<string, string>;

function assertIsCsvRow(row: unknown): asserts row is CsvRow {
  if (typeof row !== 'object'
    || Object.keys(row).length === 0
    || Object.keys(row).some((key) => typeof key !== 'string')
    || Object.values(row).some((value) => typeof value !== 'string')) {
```

Comment (bickelj) [reply src/postProposalVersions.ts:23]: The reason to check if the value is `null` first is that `typeof null` is `object`. The rest seems good to me.

Comment (jasonaowen) [reply src/postProposalVersions.ts:23]: Yes, quite right - I was focused too strongly on the TypeScript discussion we were having earlier and forgot about the JavaScript side. Good catch!

File: src/postProposalVersions.ts (hunk 2 above)
Comment (bickelj) [reply src/postProposalVersions.ts]: Added the logger. I started with my child logger setup, but then that turned out to be redundant when I'm explicitly running this one script. I suppose if it called any functions from other files it would be useful again. For now I'm skipping that step.

**Review by jasonaowen — APPROVED** [review]
Looks good! Thanks for iterating on this, @bickelj.

**Comment by bickelj** [comment]
Thanks for helping make it better!
